src/Task/Runner/TaskRunner.php

```diff
@@ -59,68 +70,72 @@ public function __construct(
      */
     public function runTasks()
     {
-        $executions = $this->taskExecutionRepository->findScheduled();
-
-        foreach ($executions as $execution) {
-            // this find is neccesary if the storage because the storage could be
-            // invalid (clear in doctrine) after handling an execution.
-            $execution = $this->taskExecutionRepository->findByUuid($execution->getUuid());
-
-            $start = microtime(true);
-            $execution->setStartTime(new \DateTime());
-            $execution->setStatus(TaskStatus::RUNNING);
-            $this->taskExecutionRepository->save($execution);
-
-            try {
-                $result = $this->handle($execution);
-
-                // this find is neccesary if the storage because the storage could be
-                // invalid (clear in doctrine) after handling an execution.
-                $execution = $this->taskExecutionRepository->findByUuid($execution->getUuid());
-                $execution->setStatus(TaskStatus::COMPLETED);
-                $execution->setResult($result);
-
-                $this->eventDispatcher->dispatch(
-                    Events::TASK_PASSED,
-                    new TaskExecutionEvent($execution->getTask(), $execution)
-                );
-            } catch (\Exception $ex) {
-                // this find is neccesary if the storage because the storage could be
-                // invalid (clear in doctrine) after handling an execution.
-                $execution = $this->taskExecutionRepository->findByUuid($execution->getUuid());
-                $execution->setException($ex->__toString());
-                $execution->setStatus(TaskStatus::FAILED);
-
-                $this->eventDispatcher->dispatch(
-                    Events::TASK_FAILED,
-                    new TaskExecutionEvent($execution->getTask(), $execution)
-                );
+        $runTime = new \DateTime();
+
+        while ($execution = $this->taskExecutionRepository->findNextScheduled($runTime)) {
```

Comment:
No here is the performance not so important. More important is that the execution is always clean and sync with the database. as you see the implementation before has loaded the whole execution which has to be executed. But in the loop, it was always refreshed because the task could reset (or the listener in the bundle) doctrine. So I change it to this more explicit implementation.